text/3888-const-self-fields.md

```diff
@@ -0,0 +1,319 @@
+# Drawbacks
+[drawbacks]: #drawbacks
+
+1. Programmers must distinguish:
+   * Fields (expr.field),
+   * Associated consts (T::CONST),
+   * And const fields (expr.METADATA).
+2. Vtable layout grows to include inline metadata, which:
+   *  Increases vtable size when heavily used.
+   * Needs careful specification for any future stable trait-object ABI.
+3. Dot syntax now covers both per-instance fields and per-type metadata; tools and docs will need to present these clearly to avoid confusion.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+### Why this design?
+* Explicitly value-only access (expr.NAME) keeps the mental model simple, as it functions similarly to a field access
+
+* If you have a trait object, you can read its per-impl metadata.
+
+* If you just have a type, associated consts remain the right tool.
+
+* By forbidding `T::NAME`, we avoid:
+  * Confusion over `dyn Trait::NAME`.
+  * Having to explain when a const is “type-level” vs “metadata-level” under the same syntax.
+* A metadata load is cheaper and more predictable than a virtual method call. Especially important when touching many trait objects in tight loops.
+
+### Why not a macro/library?
+A library or macro cannot extend the vtable layout or teach the optimizer that certain values are metadata; it can only generate more methods or global lookup tables. `const self`/`static const self` requires language and compiler support to achieve the desired ergonomics and performance.
+
+### Alternatives
+Keep using methods:
+```rust
+fn value(&self) -> u32; // remains the standard way.
+```
+Downsides:
+* Conceptual mismatch (constant-as-method).
+* Extra indirection and call overhead.
+
+# Prior art
+[prior-art]: #prior-art
+
+As of the day this RFC was published, there is no mainstream language with a similar feature. The common workaround is having a virtual function return the literal, but that does not mean we should not strive for a more efficient method.
+
+This RFC can be seen as:
+* Making explicit a pattern that compiler and runtimes already rely on internally (metadata attached to vtables).
+* Exposing it in a controlled, ergonomic way for user code.
+# Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* Is there a better declaration syntax than `static const self NAME : Type`/`const self NAME : Type`?
+* Is `obj.METADATA_FIELD` syntax too conflicting with `obj.normal_field`?
+* Is `obj.(Trait.METADATA_FIELD)` a good syntax for disambiguating?
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+* Faster type matching than `dyn Any`: Since `dyn Any` does a virtual call to get the `TypeId`, using `static const self` to store the `TypeId` would be a much more efficient way to downcast.
```

Comment:
It would be a bit more efficient. I don't think the perf difference would be all that much on modern cpus, especially when you also take the typeid comparison itself into account.